src/batSet.ml

```diff
@@ -704,24 +704,26 @@ module Concrete = struct
     
   let of_seq cmp s =
     add_seq cmp s empty
-    
-  let rec to_seq m =
-    fun () ->
-    match m with
-    | Empty -> BatSeq.Nil
-    | Node(l, v, r, _) ->
-       BatSeq.append (to_seq l) (fun () -> BatSeq.Cons (v, to_seq r)) ()
-    
-  let rec to_seq_from cmp k m =
-    fun () ->
+
+  let rec to_seq_hlp m =
     match m with
-    | Empty -> BatSeq.Nil
-    | Node(l, v, r, _) ->
-       if cmp k v <= 0 then
-         BatSeq.append (to_seq_from cmp k l) (fun () -> BatSeq.Cons (v, to_seq r)) ()
-       else
-         to_seq_from cmp k r ()
-  
+    | E -> BatSeq.Nil
+    | C(k, r, e) ->
+       BatSeq.Cons (k, fun () -> to_seq_hlp (cons_iter r e))
+      
+  let to_seq m () =
+    to_seq_hlp (cons_iter m E)
+
+  let to_seq_from cmp k m () =
+    let rec cons_enum_from k2 m e =
+      match m with
+      | Empty -> e
+      | Node (l, k, r, _) ->
+         if cmp k2 k <= 0
+         then cons_enum_from k2 l (C (k, r, e))
+         else cons_enum_from k2 r e in
+    to_seq_hlp (cons_enum_from k m E)
+
```

Comment:
Now we have two comparable implementations of `to_seq` and `to_seq_from`, this one and the one from #1011. Which one do we want to merge and reuse in the other modules?

My preference would go for merging @jabot2's code (it's always nicer to get code from fresh new contributors than grumpy old ones), but there are aspects of the implementation in #1011 that are better:
- I prefer defining `cons_enum_from` as a toplevel function, close to `cons_iter` (... which suggests that it should be called `cons_iter_from`), because it could be reused for other iterators later.
- currently `to_seq_hlp` returns a `'a BatSeq.node` and the caller takes an additional `()` argument, it would be more efficient if it returned a `'a BatSeq.t` directly. (The problem here is that OCaml assumes that `to_seq_from` is a 4-argument function, which get optimized differently from 3-argument functions.)

(I have a small preference for my naming choice, `seq_of_iter` rather than `to_seq_hlp`, but this is not very important and we could go with your choice.)

Reply:
(The implementation was rewritten, taking all these points into account, so it is now perfectly fine.)